cpp/include/cudf/strings/detail/strings_children_ex.cuh

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/detail/offsets_iterator_factory.cuh>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/utilities/default_stream.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include <stdexcept>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+namespace experimental {
+
+/**
+ * @brief Creates child offsets and chars data by applying the template function that
+ * can be used for computing the output size of each string as well as create the output
+ *
+ * The `size_and_exec_fn` is expected to be functor with 3 settable member variables.
+ * @code{.cpp}
+ * struct size_and_exec_fn {
+ *   size_type* d_sizes;
+ *   char* d_chars{};
+ *   cudf::detail::input_offsetalator d_offsets;
+ *   __device__ void operator()(size_type thread_idx) {
+ *      // functor-specific logic to resolve out_idx from thread_idx
+ *      if( !d_chars ) {
+ *         d_sizes[out_idx] = output_size;
+ *      } else {
+ *         auto d_output = d_chars + d_offsets[out_idx];
+ *         // write characters to d_output
+ *      }
+ *   }
+ * };
+ * @endcode
+ *
+ * @tparam SizeAndExecuteFunction Function must accept a row index.
+ *         It must also have member `d_sizes` to hold computed row output sizes on the 1st pass
+ *         and members `d_offsets` and `d_chars` for the 2nd pass to resolve the output memory
+ *         location for each row.
+ *
+ * @param size_and_exec_fn This is called twice. Once for the output size of each string
+ *        and once again to fill in the memory pointed to by d_chars.
+ * @param exec_size Number of threads for executing the `size_and_exec_fn` function
```

Comment:
Is this parameter `exec_size` required?
`Number of threads` may be misleading. `thrust::for_each_n` may not be launching `exec_size` threads. 

Also, what happens if `exec_size != strings_count` ?

Reply:
This is a feature that is used in a few places where the number of threads != number of output rows.
One of these places is in `cudf::strings::repeat_strings` which is [included in this PR](https://github.com/rapidsai/cudf/pull/15363/files#diff-223c9bc43a70187cc73b202a82dbe92fe0deb1ac772a24771c33fda93e3ecf3cR163-R164).